contracts/trade-shield-contract/src/action/sudo/process_orders.rs

```diff
@@ -112,33 +93,70 @@ pub fn process_orders(
             orders_to_process,
             &mut submsgs,
             env.contract.address.as_str(),
-            &mut reply_info_id,
+            reply_info_id,
             deps.storage,
             deps.querier,
         )?;
     }
 
-    for (key, order_ids) in perpetual_orders.iter() {
-        if n_perpetual_order == Some(0) {
+    Ok((submsgs, bank_msgs))
+}
+
+fn generate_perpetual_orders_submsgs(
+    deps: &mut DepsMut<ElysQuery>,
+    env: &Env,
+    n_perpetual_order: &mut Option<u128>,
+    reply_info_id: &mut u64,
+) -> Result<Vec<SubMsg<ElysMsg>>, ContractError> {
+    let querier = ElysQuerier::new(&deps.querier);
+    let mut submsgs: Vec<SubMsg<ElysMsg>> = vec![];
+
+    let perpetual_orders: Vec<(
+        String,
+        Vec<u64>,
+        (PerpetualPosition, PerpetualOrderType, String, String),
+    )> = SORTED_PENDING_PERPETUAL_ORDER
+        .prefix_range(deps.storage, None, None, Order::Ascending)
+        .filter_map(
+            |res| -> Option<(
+                String,
+                Vec<u64>,
+                (PerpetualPosition, PerpetualOrderType, String, String),
+            )> {
+                let (key, order_ids) = res.ok()?;
+                let (order_position, order_type, base_denom, quote_denom) =
+                    PerpetualOrder::from_key(&key).ok()?;
+                Some((
+                    key,
+                    order_ids,
+                    (order_position, order_type, base_denom, quote_denom),
+                ))
+            },
+        )
+        .collect();
+
+    for (key, order_ids, (order_position, order_type, base_denom, quote_denom)) in
+        perpetual_orders.iter()
+    {
+        if n_perpetual_order == &Some(0) {
             break;
         }
-        let (order_position_type, order_type, base_denom, quote_denom) =
-            PerpetualOrder::from_key(key.as_str())?;
-
         //get the price in usdc
-        let market_price =
-            match querier.get_asset_price_from_denom_in_to_denom_out(&quote_denom, &base_denom) {
-                Ok(market_price) => market_price,
-                Err(_) => {
-                    cancel_perpetual_orders(deps.storage, key, &order_ids, None)?;
-                    continue;
-                }
-            };
+        let market_price = match querier
+            .get_asset_price_from_denom_in_to_denom_out(quote_denom, base_denom)
+            .ok()
```

Comment:
I am not sure about this change, shouldn't we handle error ? what if a error occurs, we should cancel the order, right ?

Reply:
okay, so if there is an error we will cancel the order. Makes sense

Reply:
also just to confirm there are no cases where we need to fail this function in case of error ?